src/icalendar/cal/component.py

```diff
@@ -116,7 +145,21 @@ def _encode(cls, name, value, parameters=None, encode=1):
             # Don't encode already encoded values.
             obj = value
         else:
-            klass = cls.types_factory.for_property(name)
+            # Extract VALUE parameter if present, or infer it from the Python type
+            value_param = None
+            if parameters and "VALUE" in parameters:
+                value_param = parameters["VALUE"]
+            elif not isinstance(value, cls.types_factory.all_types):
+                inferred = _infer_value_type(value)
```

Comment:
If _infer_value_type is used only once and that is here, it should be located in the class. That would increase [cohesion](https://ducmanhphan.github.io/2019-03-23-Coupling-and-Cohension-in-OOP/).